### PR TITLE
Add Rule IncludeFunction

### PR DESCRIPTION
**Pull request by smnandre**
Attempt to implement #140 

Open to any feedback :)

### CODE REVIEW

File: src/Rules/Function/IncludeFunctionRule.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+{
+    protected function process(int $tokenPosition, array $tokens): void
+    {
+        $token = $tokens[$tokenPosition];
+        if (!$this->isTokenMatching($token, Token::BLOCK_NAME_TYPE, 'include')) {
+            return;
+        }
+
+        $openingTag = $this->findPrevious(Token::BLOCK_START_TYPE, $tokens, $tokenPosition);
+        Assert::notFalse($openingTag, 'Opening tag cannot not be null.');
```
Comment (VincentLanglet) [inline src/Rules/Function/IncludeFunctionRule.php:24]:
```suggestion
        Assert::notFalse($openingTag, 'Opening tag cannot be null.');
```

File: src/Rules/Function/IncludeFunctionRule.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+        $closingTag = $this->findNext(Token::BLOCK_END_TYPE, $tokens, $tokenPosition);
+        Assert::notFalse($closingTag, 'Closing tag cannot not be null.');
```
Comment (VincentLanglet) [inline src/Rules/Function/IncludeFunctionRule.php:27]:
```suggestion
        Assert::notFalse($closingTag, 'Closing tag cannot be null.');
```

File: src/Rules/Function/IncludeFunctionRule.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+final class IncludeFunctionRule extends AbstractFixableRule
+{
+    protected function process(int $tokenPosition, array $tokens): void
+    {
+        $token = $tokens[$tokenPosition];
+        if (!$this->isTokenMatching($token, Token::BLOCK_NAME_TYPE, 'include')) {
+            return;
+        }
+
+        $openingTag = $this->findPrevious(Token::BLOCK_START_TYPE, $tokens, $tokenPosition);
```
Comment (VincentLanglet) [inline src/Rules/Function/IncludeFunctionRule.php:23]: You can look for the opening tag and ending tag after the
```
if (null === $fixer) {
            return;
        }
```

check.

File: src/Rules/Function/IncludeFunctionRule.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            'Include function must be used instead of include tag.',
+            $token
+        );
+
+        if (null === $fixer) {
+            return;
+        }
+
+        $fixer->beginChangeSet();
+        $fixer->replaceToken($openingTag, '{{');
```
Comment (VincentLanglet) [inline src/Rules/Function/IncludeFunctionRule.php:39]: You're changing `{%` to `{{`.
What if it was `{%~` or `{%-` ? (This will need to be tested)

File: src/Rules/Function/IncludeFunctionRule.php [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                        break;
+                    case 'missing':
+                        $fixer->replaceToken($position, '');
+                        break;
+                    default:
+                        break;
+                }
+            }
+        }
+        $endInclude = ') }}';
```
Comment (VincentLanglet) [inline src/Rules/Function/IncludeFunctionRule.php:69]: Same, we will need to support `~%}` and `-%}`

File: src/Rules/Function/IncludeFunctionRule.php [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        if ($withoutContext) {
+            $endInclude = ', with_context = false'.$endInclude;
+        }
+        if ($ignoreMissing) {
+            $endInclude = ', ignore_missing = true'.$endInclude;
+        }
```
Comment (VincentLanglet) [inline src/Rules/Function/IncludeFunctionRule.php:75]: I'm not sure about named parameters. I didn't see them often used in twig.

So I wonder if we shouldn't try to have
```
include('template.html' , vars  , false)
```
rather than
```
include('template.html' , vars  , with_context = false)
```
and same for others options.

File: tests/Rules/Function/IncludeFunction/IncludeFunctionRuleTest.fixed.twig [hunk 7]
```diff
@@ -0,0 +1,14 @@
+{{ include('template.html' ) }}
+{{ include(['template_a.html', 'template_b.html']   , {'foo': 'bar'}  , ignore_missing = true, with_context = false) }}
+{{ include('template.html'  , with_context = false) }}
+{{ include('template.html' , vars ) }}
+{{ include('template.html' , vars  , with_context = false) }}
+{{ include('template.html' , {'foo': 'bar'} ) }}
+{{ include('template.html' , {'foo': 'bar'}  , with_context = false) }}
```
Comment (VincentLanglet) [inline tests/Rules/Function/IncludeFunction/IncludeFunctionRuleTest.fixed.twig:7]: You started to handled unnecessary whitespace with your 
```
if ($this->isTokenMatching($tokens[$tokenPosition + 1], Token::WHITESPACE_TOKENS)) {
            $fixer->replaceToken($tokenPosition + 1, '');
        }
```
check, but we still have unnecessary one here, before the `,`.

Imho, we should either
- complicate the rule to handle correctly ALL the whitespace
- or handle none

File: src/Rules/Function/IncludeFunctionRule.php [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        $fixer->replaceToken($openingTag, '{{');
+        $fixer->replaceToken($tokenPosition, 'include(');
+        if ($this->isTokenMatching($tokens[$tokenPosition + 1], Token::WHITESPACE_TOKENS)) {
+            $fixer->replaceToken($tokenPosition + 1, '');
+        }
+        $ignoreMissing = false;
+        $withoutContext = false;
+        foreach (range($tokenPosition, $closingTag - 1) as $position) {
+            $token = $tokens[$position];
+            if (Token::BLOCK_NAME_TYPE === $token->getType() || Token::NAME_TYPE === $token->getType()) {
```
Comment (VincentLanglet) [inline src/Rules/Function/IncludeFunctionRule.php:48]: Why do you need both the block name type and the name type ?
Can 4 of this keyword be found in the 2 differents types (or should we split the switch/case) ?

File: src/Rules/Function/IncludeFunctionRule.php (hunk 2 above)
Comment (smnandre) [reply src/Rules/Function/IncludeFunctionRule.php:27]: Thank you

File: src/Rules/Function/IncludeFunctionRule.php (hunk 4 above)
Comment (smnandre) [reply src/Rules/Function/IncludeFunctionRule.php:39]: Thx i keep forgetting about those

File: tests/Rules/Function/IncludeFunction/IncludeFunctionRuleTest.fixed.twig (hunk 7 above)
Comment (smnandre) [reply tests/Rules/Function/IncludeFunction/IncludeFunctionRuleTest.fixed.twig:7]: I'd say none, even if it's a bit weird-looking in the "fixed.twig" files... (i went back and forth on this when i coded)

Do you have any preference ?

File: src/Rules/Function/IncludeFunctionRule.php (hunk 6 above)
Comment (smnandre) [reply src/Rules/Function/IncludeFunctionRule.php:75]: To be honest I have absolutely no idea wich of those two args (context / missing) is the first one / second one .. 

Tried to follow "blindly" the doc : https://twig.symfony.com/doc/3.x/functions/include.html 

I remove them and it can be the goal of another rule i guess later

Comment (smnandre) [reply src/Rules/Function/IncludeFunctionRule.php:75]: Oh well..   we  generate "`include(...,  false, true)`" if there was only "`ignore_missing`" ? :| 

```php
function twig_include(Environment $env, $context, $template, $variables = [], $withContext = true, $ignoreMissing = false, $sandboxed = false)
{
```

File: src/Rules/Function/IncludeFunctionRule.php (hunk 8 above)
Comment (smnandre) [reply src/Rules/Function/IncludeFunctionRule.php]: First TYPE was useless. But we need to check both name and type to avoid false positive with vars or constants :/

File: tests/Rules/Function/IncludeFunction/IncludeFunctionRuleTest.fixed.twig (hunk 7 above)
Comment (smnandre) [reply tests/Rules/Function/IncludeFunction/IncludeFunctionRuleTest.fixed.twig]: (removed from now, tell me if you want otherwise)

File: src/Rules/Function/IncludeFunctionRule.php (hunk 6 above)
Comment (VincentLanglet) [reply src/Rules/Function/IncludeFunctionRule.php]: Yes, I had in mind to generate

Include ($template, [], false, true)

If the only option passed was "ignore missing".

I agree that "false" is unnecessary passed but I feel like people are not used to named parameter yet.

File: tests/Rules/Function/IncludeFunction/IncludeFunctionRuleTest.fixed.twig (hunk 7 above)
Comment (VincentLanglet) [reply tests/Rules/Function/IncludeFunction/IncludeFunctionRuleTest.fixed.twig]: None seems the simpler.

We can still iterate on it later if someone complains about it.

Our recommendation would to use both the include fixer with all the other spacing fixer

File: src/Rules/Function/IncludeFunctionRule.php (hunk 6 above)
Comment (smnandre) [reply src/Rules/Function/IncludeFunctionRule.php]: Done !

**Review by VincentLanglet — APPROVED** [review]

File: src/Rules/Function/IncludeFunctionRule.php [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+                        $fixer->replaceToken($position, '');
+                        break;
+                    default:
+                        break;
+                }
+            }
+        }
+        $endInclude = ') '.str_replace('%}', '}}', $tokens[$closingTag]->getValue());
+        if ($ignoreMissing) {
+            $endInclude = sprintf(', %s', $ignoreMissing ? 'true' : 'false').$endInclude;
```
Comment (VincentLanglet) [inline src/Rules/Function/IncludeFunctionRule.php:70]: The ternary condition here is always true

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

File: src/Rules/Function/IncludeFunctionRule.php (hunk 9 above)
Comment (smnandre) [reply src/Rules/Function/IncludeFunctionRule.php:70]: 😶‍🌫️

Comment (smnandre) [reply src/Rules/Function/IncludeFunctionRule.php]: Fixed, sorry

**Review by VincentLanglet — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Thanks !

File: tests/Rules/Function/IncludeFunction/IncludeFunctionRuleTest.twig [hunk 10]
```diff
@@ -0,0 +1,16 @@
+{% include 'template.html' %}
+{% include ['template_a.html', 'template_b.html'] ignore missing with {'foo': 'bar'} only %}
+{% include 'template.html' only %}
```
Comment (VincentLanglet) [inline tests/Rules/Function/IncludeFunction/IncludeFunctionRuleTest.twig:3]: @smnandre This is fixed in `{{ include( 'template.html'  , false) }}`.

I think it's wrong, and should be `{{ include( 'template.html', [], false) }}`
Not sure if it's easy without named parameters...

Comment (VincentLanglet) [reply tests/Rules/Function/IncludeFunction/IncludeFunctionRuleTest.twig:3]: Solved in https://github.com/VincentLanglet/Twig-CS-Fixer/pull/178

**Comment by 94noni** [comment]
Waw nice one @smnandre and @VincentLanglet 👌🏼

File: tests/Rules/Function/IncludeFunction/IncludeFunctionRuleTest.twig (hunk 10 above)
Comment (smnandre) [reply tests/Rules/Function/IncludeFunction/IncludeFunctionRuleTest.twig:3]: Damn really sorry to let so many mistakes :(
